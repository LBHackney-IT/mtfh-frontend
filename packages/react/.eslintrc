{
  "extends": "@hackney/eslint-config",
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/*/rollup.config.js",
          "**/*.test.ts",
          "**/*.test.tsx"
        ]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ],
    "jsx-a11y/label-has-associated-control": "off"
    // "jsx-a11y/label-has-associated-control": [
    //   2,
    //   {
    //     "labelComponents": ["Label"],
    //     "labelAttributes": ["label"],
    //     "controlComponents": ["Input"],
    //     "assert": "htmlFor",
    //     "depth": 3
    //   }
    // ]
  }
}
