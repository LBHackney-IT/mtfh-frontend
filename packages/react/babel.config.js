const babel = require("../../babel.config")

module.exports = babel
