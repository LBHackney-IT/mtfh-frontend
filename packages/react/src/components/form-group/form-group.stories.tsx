import React from "react";

import { Input } from "../input";
import { FormGroup } from "./form-group";

export default {
  title: "Form Group",
  component: FormGroup,
};

export const Base = () => {
  return (
    <FormGroup id="name" label="First Name">
      <Input />
    </FormGroup>
  );
};

export const FormGroupWithHint = () => {
  return (
    <FormGroup
      id="name"
      label="First Name"
      hint="You can find this name on your passport"
    >
      <Input />
    </FormGroup>
  );
};

export const FormGroupWithRequired = () => {
  return (
    <FormGroup id="name" label="First Name" required>
      <Input />
    </FormGroup>
  );
};

export const FormGroupWithError = () => {
  return (
    <FormGroup
      id="name"
      label="First Name"
      error="First Name is required"
      required
    >
      <Input />
    </FormGroup>
  );
};
