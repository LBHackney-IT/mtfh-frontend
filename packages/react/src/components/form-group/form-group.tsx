import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  useMemo,
} from "react"
import type * as Polymorphic from "@radix-ui/react-polymorphic"
import classNames from "classnames"

import { widthOverrides } from "../../utils"
import { ErrorMessage } from "../error-message"
import { Hint } from "../hint"
import { Label } from "../label"
import styles from "./styles.module.scss"

export interface FormGroupProps {
  id: string
  label?: string
  name?: string
  hint?: string
  error?: string | false
  required?: boolean
  children: ReactElement
  override?: number
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<
  "div",
  FormGroupProps
>

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    styles.govukFormGroup,
    {
      [styles.govukFormGroupError]: !!error,
    },
    styles.lbhFormGroup,
    widthOverrides(override),
    className,
  )

  const describedBy = useMemo(() => {
    const ids: string[] = []
    if (hint) {
      ids.push(`${id}-hint`)
    }
    if (error) {
      ids.push(`${id}-error`)
    }
    return ids.join(" ") || undefined
  }, [id, hint, error])

  const labelComp =
    typeof FormGroupComp === "string" &&
    (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label"

  return (
    <FormGroupComp
      ref={ref}
      id={id}
      className={formGroupClasses}
      aria-describedby={labelComp === "legend" ? describedBy : undefined}
      {...props}
    >
      {!!label && (
        <Label
          as={labelComp}
          htmlFor={labelComp === "label" ? `${id}-field` : undefined}
        >
          {label}
          {required ? <sup aria-hidden="true">*</sup> : ""}
        </Label>
      )}
      {!!hint && <Hint id={`${id}-hint`}>{hint}</Hint>}
      {!!error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": labelComp === "label" ? describedBy : undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  )
})
