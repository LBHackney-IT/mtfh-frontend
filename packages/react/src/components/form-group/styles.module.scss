@import "~lbh-frontend/lbh/base";
@import "~govuk-frontend/govuk/objects/form-group";
@import "~lbh-frontend/lbh/objects/form-group";

.lbh-form-group .govuk-label sup {
  color: lbh-colour("lbh-error");
}
