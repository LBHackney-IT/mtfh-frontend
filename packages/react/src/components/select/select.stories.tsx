import React from "react";

import { Select } from "./select";

export default {
  title: "Select",
  component: Select,
};

export const Base = () => {
  return (
    <Select>
      <option>Option 1</option>
      <option>Option 2</option>
      <option>Option 3</option>
    </Select>
  );
};

export const SelectWithError = () => {
  return (
    <Select error>
      <option>Option 1</option>
      <option>Option 2</option>
      <option>Option 3</option>
    </Select>
  );
};

export const SelectWithOverride = () => {
  return (
    <Select override={1 / 2}>
      <option>Option 1</option>
      <option>Option 2</option>
      <option>Option 3</option>
    </Select>
  );
};
