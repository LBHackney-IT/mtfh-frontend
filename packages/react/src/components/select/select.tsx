import React, { ComponentPropsWithoutRef, forwardRef } from "react";
import classNames from "classnames";

import { widthOverrides } from "../../utils";
import styles from "./styles.module.scss";

export interface SelectProps extends ComponentPropsWithoutRef<"select"> {
  error?: boolean;
  override?: number;
}

export const Select = forwardRef<HTMLSelectElement, SelectProps>(
  function Select({ error, className, override, ...props }, ref) {
    const selectClasses = classNames(
      styles.govukSelect,
      styles.lbhSelect,
      { [styles.govukSelectError]: error },
      widthOverrides(override),
      className
    );
    return <select ref={ref} className={selectClasses} {...props} />;
  }
);
