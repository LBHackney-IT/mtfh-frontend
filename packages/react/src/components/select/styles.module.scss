@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-select/select";
