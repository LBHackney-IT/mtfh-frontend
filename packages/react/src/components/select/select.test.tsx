import React from "react";
import { testA11y } from "@hackney/mtfh-test-utils";
import { render } from "@testing-library/react";

import { FormGroup } from "../form-group";
import { Select } from "./select";

test("it passes a11y", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="Select Test">
      <Select>
        <option value="name">Name</option>
      </Select>
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", () => {
  const { container } = render(<Select />);
  expect(container).toMatchSnapshot();
});

test("it renders correctly with error", () => {
  const { container } = render(<Select error />);
  expect(container).toMatchSnapshot();
});
