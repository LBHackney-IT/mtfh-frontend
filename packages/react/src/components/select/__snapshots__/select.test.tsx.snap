// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y 1`] = `
<div>
  <div
    class="govukFormGroup lbhFormGroup"
    id="test"
  >
    <label
      class="govukLabel lbhLabel"
      for="test-field"
    >
      Select Test
      
    </label>
    <select
      class="govukSelect lbhSelect"
      id="test-field"
      name="test"
    >
      <option
        value="name"
      >
        Name
      </option>
    </select>
  </div>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <select
    class="govukSelect lbhSelect"
  />
</div>
`;

exports[`it renders correctly with error 1`] = `
<div>
  <select
    class="govukSelect lbhSelect govukSelectError"
  />
</div>
`;
